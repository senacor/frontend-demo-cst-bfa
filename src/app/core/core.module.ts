import { NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { CcVerificationService } from './services/cc-verification.service';

@NgModule({
  declarations: [],
  imports: [
    CommonModule,
    HttpClientModule,
  ],
  exports: [],
  providers: [
    CcVerificationService,
  ],
})
export class CoreModule {
  constructor(@Optional() @SkipSelf() core:CoreModule) {
    if (core) {
      throw new Error('You should import core module only in the root module!');
    }
  }
}
