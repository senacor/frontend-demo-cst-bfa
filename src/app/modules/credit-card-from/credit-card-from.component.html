<div class="mx-auto max-w-md w-full p-6 bg-white border border-gray-200 rounded-lg shadow sm:p-6 md:p-8 dark:bg-gray-800 dark:border-gray-700">

<form [formGroup]="creditCardForm">

  <div class="">

    <div>Credit Card Information</div>

    <div class="cc-types">
      <img class="cc-types__img cc-types__img--amex">
      <img class="cc-types__img cc-types__img--visa">
      <img class="cc-types__img cc-types__img--mastercard">
      <img class="cc-types__img cc-types__img--disc">
      <img class="cc-types__img cc-types__img--generic">
    </div>

    <div class="mb-6">
      <label for="cc-holder-input" class="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
        Name on card
      </label>
      <input
        type="text"
        id="cc-holder-input"
        maxlength="60"
        class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
        autocomplete="cc-name"
        x-autocompletetype="cc-name"
        formControlName="creditCardHolder"
        #ccHolderInput
      >
    </div>

    <div class="mb-6">
      <label for="cc-number-input" class="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
        Credit card number
      </label>
      <input
        type="text"
        id="cc-number-input"
        maxlength="19"
        class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
        autocomplete="cc-number"
        x-autocompletetype="cc-number"
        formControlName="creditCardNumber"
        #ccNumberInput
      >
    </div>

    <div class="mb-6">
      <label for="cc-expiry-input" class="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
        Credit card expiration
      </label>
      <input
        type="text"
        maxlength="5"
        id="cc-expiry-input"
        class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
        autocomplete="cc-exp"
        x-autocompletetype="cc-exp"
        placeholder="MM/YY"
        formControlName="creditCardExpiration"
        #ccExpirationInput
      >
    </div>

    <div class="mb-6">
      <label for="cc-cvc-input" class="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
        Credit card expiration
      </label>
      <input
        type="text"
        maxlength="3"
        id="cc-cvc-input"
        class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
        autocomplete="cc-csc"
        x-autocompletetype="cc-csc"
        formControlName="creditCardVerification"
        #ccCCVInput
      >
    </div>

    <button
      type="button"
      class="w-full text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
      (click)="checkAndApplyForm()"
    >
      Absenden
    </button>
  </div>


</form>
</div>
