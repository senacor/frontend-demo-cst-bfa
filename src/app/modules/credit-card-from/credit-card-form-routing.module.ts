import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { CreditCardFromComponent } from './credit-card-from.component';

const routes: Routes = [
  {
    path: '',
    component: CreditCardFromComponent,
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class CreditCardFormRoutingModule {}
