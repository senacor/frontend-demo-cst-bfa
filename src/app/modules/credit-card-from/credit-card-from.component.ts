import { AfterViewChecked, Component, ViewChild } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { Router } from '@angular/router';
import { Store } from '@ngrx/store';
import { addCreditCardInformation } from '../../core/state/actions/credit-card.actions';
import { CcVerificationService } from '../../core/services/cc-verification.service';
import { CreditCard } from '../../core/models/credit-card';

@Component({
  selector: 'app-credit-card-from',
  templateUrl: './credit-card-from.component.html',
  styleUrls: ['./credit-card-from.component.scss'],
})
export class CreditCardFromComponent implements AfterViewChecked {

  // logic for the credit card form shamelessly borrowed from: https://codepen.io/murani/pen/KyVbrp

  @ViewChild('ccHolderInput') ccHolderInput!: HTMLInputElement;

  @ViewChild('ccNumberInput') ccNumberInput!: HTMLInputElement;

  @ViewChild('ccExpirationInput') ccExpirationInput!: HTMLInputElement;

  @ViewChild('ccCCVInput') ccCCVInput!: HTMLInputElement;

  creditCardForm = new FormGroup({
    creditCardHolder: new FormControl('', [Validators.required, Validators.minLength(5)]),
    creditCardNumber: new FormControl('', [Validators.required, Validators.pattern('^\\d{4} \\d{4} \\d{4} \\d{4}$')]),
    creditCardExpiration: new FormControl('', [Validators.required, Validators.pattern('^(0[1-9]|1[0-2])\/\\d{2}$')]),
    creditCardVerification: new FormControl('', [Validators.required, Validators.minLength(3), Validators.maxLength(4)]),
  });

  errorMessage: string = '';

  constructor(
    private router: Router,
    private store: Store,
    private ccService: CcVerificationService,
  ) {
  }

  ngAfterViewChecked() {
    this.creditCardForm.get('creditCardNumber')?.valueChanges.subscribe((value) => {
      this.updateCreditCardNumberInput(value as string);
      this.highlightCC(value as string);
    });

    this.creditCardForm.get('creditCardExpiration')?.valueChanges.subscribe((value) => {
      this.updateExpiryInput(value as string);
    });
  }

  checkAndApplyForm() {
    this.errorMessage = '';
    if (this.creditCardForm.valid) {
      const creditCard: CreditCard = {
        holder: this.creditCardForm.value.creditCardHolder ?? '',
        ccNumber: this.creditCardForm.value.creditCardNumber ?? '',
        expiration: this.creditCardForm.value.creditCardExpiration ?? '',
        ccv: this.creditCardForm.value.creditCardVerification ?? '',
      };

      // call backend verification service
      this.ccService.verifyCreditCardInformation(creditCard).subscribe({
        next: (verifiedCreditCard: CreditCard) => {
          this.store.dispatch(addCreditCardInformation({ creditCard: verifiedCreditCard }));
          this.router.navigate(['/main'], { replaceUrl: true });
        },
        error: (error: Error) => {
          this.errorMessage = error.message;
        },
      });
    } else {
      this.errorMessage = 'The form is invalid, please check again';
    }
  }

  highlightCC(ccValue: string) {
    let ccCardType = '',
      ccCardTypePatterns = {
        amex: /^3/,
        visa: /^4/,
        mastercard: /^5/,
        disc: /^6/,

        generic: /(^1|^2|^7|^8|^9|^0)/,
      };

    for (const cardType in ccCardTypePatterns) {
      // @ts-ignore
      if (ccCardTypePatterns[cardType].test(ccValue) ) {
        ccCardType = cardType;
        break;
      }
    }

    let activeCC = document.querySelector('.cc-types__img--active'),
      newActiveCC = document.querySelector(`.cc-types__img--${ccCardType}`);

    if (activeCC) activeCC.classList.remove('cc-types__img--active');
    if (newActiveCC) newActiveCC.classList.add('cc-types__img--active');
  }

  updateCreditCardNumberInput(value: string) {
    let ccNumberValue = '';
    const v = value.replace(/\s+/g, '').replace(/[^0-9]/gi, '');
    const matches = v.match(/\d{4,16}/g);
    const match = matches && matches[0] || '';
    const parts = [];

    for (let i = 0, len = match.length; i < len; i += 4) {
      parts.push(match.substring(i, i + 4));
    }

    if (parts.length) {
      ccNumberValue = parts.join(' ');
    } else {
      ccNumberValue = value;
    }

    this.creditCardForm.get('creditCardNumber')?.patchValue(ccNumberValue, { emitEvent: false, onlySelf: true });
  }

  updateExpiryInput(value: string) {
    const sanitizedInput = (value as string).replace(/^\D/g, '').replace(/\./g, '').replace(/\//g, '');
    const formattedValue = sanitizedInput.replace(/^(0[1-9]|1[0-2])/, '$1/').replace(/^(0[1-9]|1[0-2])(\d{2})/, '$1/$2');
    this.creditCardForm.get('creditCardExpiration')?.patchValue(formattedValue, { emitEvent: false, onlySelf: true });
  }
}
