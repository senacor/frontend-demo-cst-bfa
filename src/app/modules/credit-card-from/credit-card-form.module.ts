import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { CreditCardFromComponent } from './credit-card-from.component';
import { CreditCardFormRoutingModule } from './credit-card-form-routing.module';
import { SharedModule } from '../../shared/shared.module';

@NgModule({
  imports: [
    CommonModule,
    SharedModule,
    FormsModule,
    ReactiveFormsModule,
    CreditCardFormRoutingModule,
  ],
  exports: [CreditCardFromComponent],
  declarations: [CreditCardFromComponent],
})
export class CreditCardFormModule {}
