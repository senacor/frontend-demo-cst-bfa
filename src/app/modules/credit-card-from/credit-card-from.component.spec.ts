import { ComponentFixture, TestBed } from '@angular/core/testing';
import { CreditCardFromComponent } from './credit-card-from.component';
import { expect } from '@jest/globals';

describe('CreditCardFromComponent', () => {
  let component: CreditCardFromComponent;
  let fixture: ComponentFixture<CreditCardFromComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ CreditCardFromComponent ],
    })
      .compileComponents();

    fixture = TestBed.createComponent(CreditCardFromComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});
