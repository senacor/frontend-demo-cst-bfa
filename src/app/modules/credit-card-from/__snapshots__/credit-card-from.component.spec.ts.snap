// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditCardFromComponent should create 1`] = `
<div
  id="root0"
>
  <div
    class="mx-auto max-w-md w-full p-6 bg-white border border-gray-200 rounded-lg shadow sm:p-6 md:p-8 dark:bg-gray-800 dark:border-gray-700"
  >
    <form
      class="ng-untouched ng-pristine ng-invalid"
      novalidate=""
    >
      <div>
        <div>
          Credit Card Information
        </div>
        <div
          class="cc-types"
        >
          <img
            class="cc-types__img cc-types__img--amex"
          />
          <img
            class="cc-types__img cc-types__img--visa"
          />
          <img
            class="cc-types__img cc-types__img--mastercard"
          />
          <img
            class="cc-types__img cc-types__img--disc"
          />
          <img
            class="cc-types__img cc-types__img--generic"
          />
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
            for="cc-holder-input"
          >
             Name on card 
          </label>
          <input
            autocomplete="cc-name"
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500 ng-untouched ng-pristine ng-invalid"
            data-cy="cc-holder-input"
            formcontrolname="creditCardHolder"
            id="cc-holder-input"
            maxlength="60"
            type="text"
            x-autocompletetype="cc-name"
          />
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
            for="cc-number-input"
          >
             Credit card number 
          </label>
          <input
            autocomplete="cc-number"
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500 ng-untouched ng-pristine ng-invalid"
            data-cy="cc-number-input"
            formcontrolname="creditCardNumber"
            id="cc-number-input"
            maxlength="19"
            type="text"
            x-autocompletetype="cc-number"
          />
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
            for="cc-expiry-input"
          >
             Credit card expiration 
          </label>
          <input
            autocomplete="cc-exp"
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500 ng-untouched ng-pristine ng-invalid"
            data-cy="cc-expiry-input"
            formcontrolname="creditCardExpiration"
            id="cc-expiry-input"
            maxlength="5"
            placeholder="MM/YY"
            type="text"
            x-autocompletetype="cc-exp"
          />
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
            for="cc-cvc-input"
          >
             CCV 
          </label>
          <input
            autocomplete="cc-csc"
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500 ng-untouched ng-pristine ng-invalid"
            data-cy="cc-cvc-input"
            formcontrolname="creditCardVerification"
            id="cc-cvc-input"
            maxlength="4"
            type="text"
            x-autocompletetype="cc-csc"
          />
        </div>
        <button
          class="w-full text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
          type="button"
        >
           Apply credit card information 
        </button>
      </div>
    </form>
  </div>
</div>
`;
